boreal/src/bitmaps.rs

```diff
@@ -0,0 +1,166 @@
+use std::ops::BitOrAssign;
+
+#[derive(Debug, Clone, Copy, Default)]
+/// A bitmap with 256 bits
+pub struct Bitmap {
+    low: u128,
+    high: u128,
+}
+
+impl Bitmap {
+    const HALF: u8 = 128;
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn mask(bit: u8) -> u128 {
+        1u128 << (bit & 127)
+    }
+
+    fn get_half(&self, bit: u8) -> u128 {
+        if bit < Self::HALF {
+            self.low
+        } else {
+            self.high
+        }
+    }
+
+    fn get_half_mut(&mut self, bit: u8) -> &mut u128 {
+        if bit < Self::HALF {
+            &mut self.low
+        } else {
+            &mut self.high
+        }
+    }
+
+    #[must_use]
+    #[inline(always)]
+    pub fn get(&self, bit: u8) -> bool {
+        let mask = Self::mask(bit);
+        let half = self.get_half(bit);
+        half & mask != 0
+    }
+
+    #[inline(always)]
+    pub fn set(&mut self, bit: u8, value: bool) {
+        let mask = Self::mask(bit);
+        let half = self.get_half_mut(bit);
+        if value {
+            *half |= mask;
+        } else {
+            *half &= !mask;
+        }
+    }
+
+    #[inline(always)]
+    pub fn invert(&mut self) {
+        self.low = !self.low;
+        self.high = !self.high;
+    }
+
+    #[inline(always)]
+    pub fn count_ones(&self) -> usize {
+        (self.low.count_ones() + self.high.count_ones()) as usize
+    }
+
+    pub fn iter(&self) -> Iter {
+        Iter(*self)
+    }
+}
+
+/// implement `|=`
+impl BitOrAssign for Bitmap {
+    fn bitor_assign(&mut self, rhs: Self) {
+        self.low |= rhs.low;
+        self.high |= rhs.high;
+    }
+}
+
+pub struct Iter(Bitmap);
+
+impl Iterator for Iter {
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This function can be a bit simpler:

```rust
let (v, offset) = if self.bitmap.low != 0 {
    (&mut self.bitmap.low, 0)
} else if self.bitmap.high != 0 {
    (&mut self.bitmap.high, 128)
} else {
    return None;
};

// Safety: this value is contained in [0; 127] so it always fits in a u8.
let t: u8 = v.trailing_zeros().try_into().unwrap();
*v &= !(1 << t);
Some(offset + t)
```

the generated assembly is reduced as well